[LeowLiYong] iP

ginloy opened the pull request:
# D
### Simple task tracking chatbot
- Free
- Lightweight
- Easy to use text-based interface
- Many features and more coming soon!

#### Installation
1. Install jar file from [here](https://github.com/ginloy/ip/releases)
2. Put jar file in empty directory
3. Double click it

#### Features
- [x] Basic tasks
- [x] Deadlines
- [x] Events
- [x] Task finding
- [x] Simple GUI
- [x] Task sorting
- [ ] Filtering by type (Coming soon)

Commit:
Add Gradle support

ginloy committed:
Completed level 1

ginloy committed:
edited .gitignore

ginloy committed:
Merge pull request #1 from ginloy/Lvl1

Lvl1

ginloy committed:
Added TaskList class

ginloy committed:
Edited engine to use TaskList and changed formatting of tasklist

ginloy committed:
Merge pull request #2 from ginloy/lvl-2

Lvl 2

ginloy committed:
added Task class

ginloy committed:
modified TaskList to use Task class

ginloy committed:
added Pair and Util classes, Engine not working yet

ginloy committed:
can now mark and unmark tasks with very minimal exception handling

ginloy committed:
Merge pull request #3 from ginloy/lvl-3

Lvl 3

ginloy committed:
added Deadline, Todo and Event classes, Engine support and parsing still not implmented

ginloy committed:
can now add 3 different types of tasks with personal flags

ginloy committed:
Added launch script to compile and run program.
Added some simple input error handling (blank/missing fields)
Automated testing with runtest.sh

ginloy committed:
modified .gitignore

ginloy committed:
modified .gitignore

ginloy committed:
Merge pull request #4 from ginloy/branch-TextUiTesting

Branch text UI testing

ginloy committed:
can now delete tasks from list

ginloy committed:
Merge pull request #5 from ginloy/branch-delete

can now delete tasks from list

ginloy committed:
Added Either monad to help with errors.

ginloy committed:
made flatmap more general

ginloy committed:
add save tasks functionality

ginloy committed:
made directory path more generalisable

ginloy committed:
Add date functionality to tasks

Deadlines and events now accepts official dates as arguments instead of random strings.

ginloy committed:
Add find by date functionality

findbydate command allows user to find and list all tasks
that contains input date.

ginloy committed:
Merge pull request #6 from ginloy/branch-Level-7

Branch level 7

ginloy committed:
Merge branch 'master' into branch-Level-8

ginloy committed:
Merge pull request #7 from ginloy/branch-Level-8

Branch level 8

ginloy committed:
Add simple monadic parser combinator class

The class allows clients to chain different parsers together to build
specific parsers similar to parsec from haskell. Result class was also added
to work with the Parser class.

ginloy committed:
Add more combinators

Added manyUntil, some and someUntil combinators.

ginloy committed:
Add extra methods

Added some more combinators and general purpose parsers.

ginloy committed:
Add match function similar to rust enums for Either class

ginloy committed:
Stop using unsafe get methods in Result

ginloy committed:
Edit Either class more general

ginloy committed:
Start organizing into packages and more classes

ginloy committed:
Sort classes into packages

Also added Storage and Ui classes for object storage and
IO respectively.

ginloy committed:
Add date parser

ginloy committed:
Add some command classes

Reorganized more code into packages. Added some more parsers and
refactored some existing parsers. Added some more utility
functions. Main function still not linked with changes.

ginloy committed:
Complete Command classes

Added unmark, delete and findbydate commands.

ginloy committed:
Add command classes

Accidentally failed to commit command classes previously.

ginloy committed:
Modify Duke to use new classes and packages

ginloy committed:
Merge pull request #8 from ginloy/branch-organize

Branch organize

ginloy committed:
Merge pull request #9 from ginloy/add-gradle-support

Add Gradle support

ginloy committed:
Add tests for Util class

ginloy committed:
Add filterOrElse method for Either class

ginloy committed:
Add some tests for Either class

ginloy committed:
Add some more tests for Either\n\nChanged toString method for Result.

Tohtoroo commented on src/main/java/Duke.java:
```diff
@@ -1,10 +1,49 @@
+import io.Storage;
+import io.Ui;
+import task.TaskList;
+
```
Consider adding JavaDoc comments?

Tohtoroo commented on src/main/java/command/Add.java:
```diff
@@ -0,0 +1,75 @@
+import task.Todo;
+import util.Parser;
+import util.Util;
+
```
Consider adding JavaDoc comment here as well!

Tohtoroo commented on src/main/java/command/Add.java:
```diff
@@ -0,0 +1,75 @@
+        return Parser.skipSpace()
+                .ignoreThen(Parser.strParserIgnoreCase("deadline"))
+                .thenIgnore(Parser.skipSpace())
+                .ignoreThen(Parser.strUntil(Parser.strParserIgnoreCase("/by")))
```
Perhaps this empty space is a mistake?

Tohtoroo commented on src/main/java/command/Delete.java:
```diff
@@ -0,0 +1,37 @@
+public class Delete implements Command {
+    private final int taskNum;
+
+    private static final String ERROR = "Task not in list.";
```
Perhaps the static attributes should be above the private final attributes according to coding standards

Tohtoroo commented on src/main/java/task/Event.java:
```diff
@@ -0,0 +1,42 @@
+public class Event extends Task {
+    private final LocalDate start;
+    private final LocalDate end;
+    private static final DateTimeFormatter DATE_FORMAT = DateTimeFormatter.ofPattern("dd MMM YY");
```
Same as above, perhaps move this static attribute above

Tohtoroo commented on src/main/java/task/Event.java:
```diff
@@ -0,0 +1,42 @@
+    public Event markNotDone() {
+        return new Event(this.desc, false, this.start, this.end);
+    }
+
```
Perhaps delete the unneeded white space

Tohtoroo reviewed (commented):
Overall, the code is very comprehensive, and the packages are done well. It is just some style issues that I'm pointing out.

runoutofit commented on src/main/java/Result.java:
```diff
@@ -42,16 +37,23 @@ public boolean isError() {
-    public T getRes() {
-        return this.result.fromLeft(null).first();
-    }
+    // public T getRes() {
```
i like that you decided not to use getters since they are bad OO design

runoutofit commented on src/main/java/Engine.java:
```diff
@@ -1,8 +1,19 @@
     private final Scanner scanner;
     private final TaskList taskList;
+    private static final String DIRECTORY = "./data";
+    private static final String FILE_NAME = "taskList.dat";
```
i like that you use upper snake case for naming constant variables

runoutofit commented on src/main/java/Engine.java:
```diff
@@ -5,52 +5,94 @@ public class Engine {
 
-    private String getOutput(Command command, String text) {
+    private String getOutput(Command command, String args) {
         switch(command) {
```
i like that you use switch case instead of if-else statements since it improves overall code readability

runoutofit reviewed (commented):
overall good code

ginloy committed:
Change jar file output name

ginloy committed:
Add JavaDoc comments for command package

ginloy committed:
Add JavaDoc comments for io package

ginloy committed:
Add JavaDoc comments for task package

ginloy committed:
Move Parser and Result into new parsing package

ginloy committed:
Add JavaDoc comments for util package

ginloy committed:
Change formatting of comments

ginloy committed:
Edit Result and Parser to use Either's filterOrElse

ginloy committed:
Add javaDoc comments for parsing package

ginloy committed:
Shift static constants above object attributes

ginloy committed:
Add find command

Can now find tasks by keywords.
Added find tests.

ginloy committed:
Modify find to work only if all keywords are matched

ginloy committed:
Merge branch 'branch-CodingStandard'
Changed some formats to satisfy coding standards.

ginloy committed:
Merge branch 'branch-JavaDoc'

Merge conflicts resolved.

ginloy committed:
Add some more documentation

ginloy committed:
Change logo

ginloy committed:
Add checkstyle

ginloy committed:
Add javadoc comments to main duke class

ginloy committed:
Edit build file and gitignore

ginloy committed:
Add Gui class

ginloy committed:
gitignore now working on .vscode

ginloy committed:
Modify gitignore

ginloy committed:
Change Ui class to interface

Shifted UI functionality to new Cli class. Gui can now implement Ui.

ginloy committed:
Gui in progress

ginloy committed:
Change CLI to have similar interface as javafx so they would be interchangeable

Also fixed a bug with Cli.split method.

ginloy committed:
Fixed bug program not creating storage directory if doesnt exist

ginloy committed:
Remove unused imports

ginloy committed:
Simple GUI working

ginloy committed:
Add clipping of pictures, padding

ginloy committed:
Use fxml and can resize window

ginloy committed:
Shift logic back to Gui class

ginloy committed:
Remove unused imports

ginloy committed:
Flip reply dialogue box

ginloy committed:
Remove unused imports

ginloy committed:
Add some assertions

Added assertions that gui properties were are not null because
respective properties were not final. They were also not intialized by
constructors by design hence I felt that additional checks were needed.

ginloy committed:
Change public properties to private

Added package private getters and setters for gui properties and
MainWindow properties so that either class can access each
other's properties directly.

ginloy committed:
Add more null assertions for non-final properties

ginloy committed:
Add more documentation

ginloy committed:
Merge pull request #11 from ginloy/branch-assertions

Branch assertions

ginloy committed:
Merge branch 'master' into branch-codeQuality

ginloy committed:
Resolve merge conflicts

ginloy committed:
Merge pull request #12 from ginloy/branch-codeQuality

Branch code quality

ginloy committed:
Change done property to isDone

ginloy committed:
Add taskList sort method

ginloy committed:
Add sort command

Bot can now sort tasks with dates by date. Fixed bug with taskList sort
method.

ginloy committed:
Improve GUI

Changed background to dark mode. Changed font color of bot replies.
Changed font color of error messages.

ginloy committed:
Refactor commands to use Ui#showError method

ginloy committed:
Refactor sort command replies into constants

ginloy committed:
Fix bug with integer parser

Was throwing exception with non integer inputs instead of returning
error message.

ginloy committed:
Fix scrollpane bug

Scrollpane was not scrolling with mousewheel until scroll bar is
clicked.

ginloy committed:
Remove blue focus outlines from gui

ginloy committed:
Change font, font size and starting window size

ginloy committed:
Remove extra brace

ginloy committed:
Reformat and resolve ide warnings

ginloy committed:
Reformat all files

ginloy committed:
Improve error messages and add help command

ginloy committed:
Change help error to have error formatting

ginloy committed:
Change unit tests asserts

ginloy committed:
Reformat code

ginloy committed:
Made help arguments optional

Help command now lists all commands if no arguments supplied.

ginloy committed:
Add user guide

ginloy committed:
Update project root README

ginloy committed:
Change application icon

ginloy committed:
Change Ui screenshot

ginloy committed:
Add installation instructions in user guide

ginloy committed:
Resolve checkstyle issues

ginloy committed:
Resolve checkstyle issues with test classes

ginloy committed:
Add border to dialogue boxes

ginloy committed:
Update Ui.png

ginloy committed:
Fix Ui.png misalignment

ginloy committed:
Move main class into its own package

ginloy committed:
Add additional installation instructions

ginloy committed:
Add more parsers and modified command to use parsers

ginloy committed:
Fix ascii art not displaying properly on windows

ginloy committed:
Shift installation instructions to realease page
